drivers: clocks: sam0: Fix devicetree usage of GCLK and MCLK

nandojve opened the pull request:
Some platforms require special clock selection options. This could be made using the already defined assigned-clocks from Linux devicetree. This can be used on Atmel SAM0 SoC to enable peripherals clocks in distinct places like PM and MCLK. Old devices had defined the peripheral clock enable bit at PM. On the newer devices this was extracted on a dedicated memory section called Master Clock (MCLK). This change excludes the dedicated bindings in favor of a generic approach that cover all cases. It uses the assigned-clocks and assigned-clock-names properties to generalize those conditions.

With this, the clocks are very flexible now and this gives the liberty to developer to customize the clock source from a generic clock or configure the direct connections. The drivers that require those properties were reworked to use the new solution. Some drivers still access direct the properties and should be reworked in the future.

Note:
1- This changes is a step necessary to have proper clock drivers for these SoCs.
2- This was splitted from #81819 and should be merged first.

depends on #82991

nandojve committed:
dts: clock: Add atmel,assigned-clock property

Some platforms require special clock selection options. This could be
made using the already defined assigned-clocks from Linux clocks.

  See https://github.com/devicetree-org/dt-schema/blob/93ee8008959e362548168eaa7bdc20c115f3d586/dtschema/schemas/clock/clock.yaml#L24

This introduces the vendor atmel,assigned-clocks and
atmel,assigned-clock-names properties to generalize those conditions
in Zephyr for Atmel sam0 SoC series.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

henrikbrixandersen commented on drivers/can/can_sam0.c:
```diff
@@ -109,6 +113,12 @@ static int can_sam0_get_core_clock(const struct device *dev, uint32_t *rate)
 {
+	*cfg->mclk |= cfg->mclk_mask;
+
+	GCLK->PCHCTRL[cfg->gclk_id].reg = GCLK_PCHCTRL_CHEN
+					| GCLK_PCHCTRL_GEN(cfg->gclk_gen)
+					;
```
Formatting looks off here?

henrikbrixandersen replied:
How is this a reason for accepting poorly formatted code? What am I missing?

nandojve replied:
I think this is not about poorly formatted code is about style. Since I need to apply the same kind of changes for the whole platform I use the same style for all. The styling always were a problem and always will be since everyone has an opinion about it and they are contradictory most of time. I can accept to change and use your preferences on the CAN driver since you maintain this drivers pretty well. I'll send a fix for this driver.

henrikbrixandersen commented on boards/atmel/sam0/samd20_xpro/samd20_xpro.yaml:
```diff
@@ -9,7 +12,6 @@ toolchain:
 flash: 256
 ram: 32
 supported:
-  - adc
```
Were these ADC changes meant to into this commit? If so, please add some details on why in the commit message.

henrikbrixandersen replied:
Sounds like something that should be submitted as a separate PR?

nandojve replied:
It can be extract on another PR, yes. I'll do it.

henrikbrixandersen commented on dts/bindings/base/base.yaml:
```diff
@@ -60,6 +60,14 @@ properties:
     type: string-array
     description: name of each clock
 
+  assigned-clocks:
+    type: phandle-array
+    description: Assigned-clock information
+
+  assigned-clock-names:
+    type: string-array
+    description: Name of each assigned-clock
```
The commit message says "already defined" properties - but these properties are new in Zephyr and not mentioned in the Devicetree Specification (v0.4).

I would like to hear @danieldegrasse opinion on the use of these and possible conflict with the upcoming clock controller rework.

decsny replied:
This looks similar to Linux clock DT description

danieldegrasse replied:
This is similar to the Linux clock bindings, which are now part of the generic DT schema: https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/clock/clock.yaml#L126.

The way Linux uses this is like so:

Peripheral devicetree definition

```
&uart0 {
    assigned-clocks = <&clock_controller UART0_CLOCK_DIV &clock_controller UART0_CLOCK_MUX>;
    /*
     * At boot time, the common clock framework will request the clock given by UART0_CLOCK_ID
     * to run at 100Mhz  via clk_set_rate()
     */
    assigned-clock-rates = <DT_FREQ_M(100)>;
    /*
     * At boot time, the common clock framework will set the parent of UART0_CLOCK_MUX to the clock
     * given by PLL_0_CLOCK_ID via a clk_set_parent() call
     */
    assigned-clock-parents = <0 &clock_controller PLL0_CLOCK_ID>;
}
```

This is just a rough sketch of the usage- the index of each value for `assigned-clock-xx` properties matches with the clock at that index in the `assigned-clocks` property.

I have deliberately avoided using `assigned-clocks` because the way we configure clocks in the clock controller rework varies so drastically from how Linux handles it.

Given the current state of the clock framework, we could possible support using `assigned-clocks`, but it would come at some additional flash cost- the main thing missing from the clock framework right now is a generic "set_parent" API call- this isn't something I'm keen to add, because increasing the size of the API structure for clocks will significantly increase the image size.

danieldegrasse replied:
Ok, after looking at this a bit more I really don't think the current usage of `assigned-clocks` property is correct here either way. The intended usage of the assigned-clocks property is to define a clock, which will then be configured at boot time based on the `assigned-clock-rates` or `assigned-clock-parents` property. Neither is present here.

Based on what I can tell from the SAM D20 reference manual (I am not familiar with this part), the gclk setting here is configuring which of the 8 generic clock generator outputs will be used for each peripheral, right? If you wanted to do this with `assigned-clocks`, you would need something like the following:

```
assigned-clocks = <&gclk GCLK_SERCOM0_CORE>;
assigned-clock-parents = <&gclk GCLKGEN0>;
```

That being said, while this is definitely an improvement it really doesn't enable *that* much more configuration, does it? From what I can tell, the SAM D20 still has the actual sources/dividers of the generic clock generators hardcoded in the SOC init: https://github.com/zephyrproject-rtos/zephyr/blob/d04b3e03232df610833e1c5f91323a8f458b071d/soc/atmel/sam0/common/soc_samd2x.c#L246

Again, I'm very much unfamiliar with the SOC- please tell me if I'm missing something here.

Personally, I'd prefer we don't continue adding disparate solutions to clock issues like these, and try to define a unified framework. That being said, this *is* still an improvement, so **provided we use assigned-clocks as I described**, I won't block it.

danieldegrasse replied:
> This is an attempt to solve a big mess at moment with this platform. It is true that it is not fully implemented. I understand that, it don't need to be fully implemented or 100% correct to start to get feedback and pass some idea. Besides that, the proposal works as expected and improve the overall situation, which is the goal here.

Sure- and I really don't want to stall progress simply because it "doesn't seem right". The goal is always to move things forward.

> IMHO, this provide a base to talk and to understand what could be the right direction. It is easy to see that what is missing could be in future added inside a clock driver or future alternative, for instance what is inside SoC with GCLK. BTW, why should I invest more time in something that I don't know if can move forward? Maybe the selection of assigned-clocks was not a good option!

Sure, I agree here. Honestly, if you want to move this forwards as is I think the way to do it is to avoid adding the `assigned-clocks` property. My core concern is that if we add the `assigned-clocks` property into Zephyr, we now are starting to expand generic code- there is a much higher bar to do that than there is to enable users to perform some clock configuration for the SAM0 SOC.

Perhaps we could just add the index of the GENCLK that a given peripheral was selecting as a second specifier on the `gclk` node? Or does that not make sense?

decsny replied:
> The goal is always to move things forward.

why is the goal to always move things forward? who told you that? if you really believe something shouldnt happen and you know its going to cause you grief later to undo it, then dont let it happen, don't sugarcoat it and make people think there's always a chance you'll let them do whatever they want. Always letting any things move forward just for the sake of pleasing people is one good way to have projects fail

>  there is a much higher bar to do that than there is to enable users to perform some clock configuration for the SAM0 SOC. Perhaps we could just add the index of the GENCLK that a given peripheral was selecting as a second specifier on the gclk node? Or does that not make sense?

I share the same opinion here, I don't know why we need to add generic code for all bindings and add clock macros that we never plan to use. The arch WG already discussed Daniel's clock RFC multiple times and so far there is no dissents, and multiple companies are committing to implement to that RFC once it is merged. So there's no reason to believe we won't be going another direction for now. Therefore adding this generic code is NACK from me

Change request: make a SAM-specific DT like all the other SOCs for now until an archWG-sanctioned unified clock framework is available.

decsny replied:
The argument for doing this type of way in this PR is of course the age old "adhere to the linux" doctrine and something about project agnostic bindings, but I am not convinced that this is really as important, if we decided that the linux way of doing it is not appropriate for the Zephyr use case, which Daniel has talked to in the past

nandojve replied:
> > This is an attempt to solve a big mess at moment with this platform. It is true that it is not fully implemented. I understand that, it don't need to be fully implemented or 100% correct to start to get feedback and pass some idea. Besides that, the proposal works as expected and improve the overall situation, which is the goal here.
> 
> Sure- and I really don't want to stall progress simply because it "doesn't seem right". The goal is always to move things forward.
> 
> > IMHO, this provide a base to talk and to understand what could be the right direction. It is easy to see that what is missing could be in future added inside a clock driver or future alternative, for instance what is inside SoC with GCLK. BTW, why should I invest more time in something that I don't know if can move forward? Maybe the selection of assigned-clocks was not a good option!
> 
> Sure, I agree here. Honestly, if you want to move this forwards as is I think the way to do it is to avoid adding the `assigned-clocks` property. My core concern is that if we add the `assigned-clocks` property into Zephyr, we now are starting to expand generic code- there is a much higher bar to do that than there is to enable users to perform some clock configuration for the SAM0 SOC.
> 
> Perhaps we could just add the index of the GENCLK that a given peripheral was selecting as a second specifier on the `gclk` node? Or does that not make sense?

Hi @danieldegrasse ,

I'm glad that you understood what is my goal and you are willing to help and guide. This is what I expect from an experienced professional with knowledge on the field.

> Honestly, if you want to move this forwards as is I think the way to do it is to avoid adding the assigned-clocks property. My core concern is that if we add the assigned-clocks property into Zephyr, we now are starting to expand generic code- there is a much higher bar to do that than there is to enable users to perform some clock configuration for the SAM0 SOC.

After this comment it is clear to me your concern which is not have the `assigned-clocks` property. I would like to talk more about the problem and forget the name "assigned-clocks" for the moment. To make it more clear let's see the below clock trees. I select 2 devices that we can explore the details. All the other will, somehow, have similar blocks.

SAME54
![image](https://github.com/user-attachments/assets/7a282183-3e30-4ff9-a87c-5d0c4c1ca54e)

SAMD21
![image](https://github.com/user-attachments/assets/9ef59c46-c1c3-47a3-84b6-02c4da4196ae)

1- I would like first point out how the clocks are enabled. In the SAME5x it is on the MCLK and for the SAMD it is on PM. When looking in deep the MCLK was extracted from PM block and put into another region on newer SoC versions. This means that a PM block still exists.

2- Those devices have a complex clock matrix which is "recursive" in terms of software. You can pick the output from one Generator and use as Input for another block which can be `sysctrl` or `oscctrl`.

3- The right part of `gclk` block maps a peripheral channel into one Generator.

4- Newer devices start to have direct connection to other blocks like `osck32ctrl`. This means that the configuration will do not have a Generator Output connected. The newer devices extracted again blocks from `sysctrl`.

The difficult that I found is that the user must configure the `gclk` as board specific. This is peripheral independent too. So, to enable a RTC the user must:

1- Enable peripheral at PM or MCLK block.
2- (left side of the `gclk` block) Define and Enable a GeneratorX with the right input clock and define the output frequency.
3- (right side of the `gclk` block) Select the GeneratorX as the clock for the peripheral.

[1] is defined inside dts/arm/atmel/soc.dtsi file. [2] and [3] are defined by project. The problem here is that it is hard to generalize because using conventional mechanism because SAME5x RTC has a direct connection with `osck32ctrl` block.

I could generalize the solution to the use on a way that it does not matter if you have a PM or a MCLK block. The macros will do the job to select the correct bit to enable/disable the peripheral. This is a requirement to have a proper clock driver on my view, I hope I'm wrong here. When I use an indirection inspired by `assigned-clocks` I could select the correct clock source and perform the correct configuration independent of the architecture using a macro to capture the correct information and store on the driver config part. This can be validated using the `clocks` lines. This was the trick that generalize the problem and I believe it is still to implement the clock driver. Again, I hope I'm not lost here.

So, my question is do you see an alternative to the indirection ? I know that NXP have very complex clocks and I was wondering if you have similar problems that you could show to me how did you solve.

About `assigned-clocks` now. This was the source of inspiration that provide the indirection way to configure the clocks for this platform. That solved the issue to have many different clock implementations and configurations reducing the development to a minimal. I could generalize everything mostly with a simple macro. Maybe `assigned-clocks` is not a good option for Zephyr but I start to think that once we start to have very complex devices in tree we need, maybe, this type the solution.

I'm willing to use a conventional solution if it can provide the similar level of flexibility. So, do you see something that I can use as an alternative to do this type of configuration?

Important: PLL can use a "right side" of a Generator as input and as consequence the PLL can be an Input for another Generator.

Alternative to `assigned-clocks`. Do you see any problem if I continue to use the indirection but define a private vendor binding ?

> Perhaps we could just add the index of the GENCLK that a given peripheral was selecting as a second specifier on the gclk node? Or does that not make sense?

I'm not sure if I understood you suggestion here. Could you draw an example that fits for both SoC using the same solution?

nandojve replied:
Hi @danieldegrasse ,

I was wondering if you had time to check my last comments.

nandojve replied:
Hi @danieldegrasse ,

Thank you for your incites. I defined a private vendor binding atmel,assigned-clocks. I think it is better keep this way and wait #72102.

nandojve replied:
One thing that still not clear to me yet:

Defined at SoC dtsi:

```
#define SERCOM0_GCLK_ID 34 // GCLK ID for SERCOM0
#define SERCOM_GEN0_ID 0 // ID to select generator 0
#define SERCOM_GEN0_ID 1 // ID to select generator 1

&sercom0 {
	clocks = <&gclk SERCOM0_GCLK_ID SERCOM_GEN0_ID>, <&mclk OFFSET BIT>;
};
```

Can user only redefine one entry, like below, on their board and still have a valid **<&mclk OFFSET BIT>** available?
```
[1]
&sercom0 {
	clocks = <&gclk SERCOM0_GCLK_ID SERCOM_GEN1_ID>;
};
```

or he needs to define all the entries:
```
[2]
&sercom0 {
	clocks = <&gclk SERCOM0_GCLK_ID SERCOM_GEN1_ID>, <&mclk OFFSET BIT>;
};
```

If developer can override using [1] the **atmel,assigned-clocks** may not be necessary. Besides it is very easy to make mistakes with copy and paste and hard to find the issue then.

henrikbrixandersen replied:
> Can user only redefine one entry, like below, on their board and still have a valid <&mclk OFFSET BIT> available?

Sadly, no. That will not work, as the entire property value is overwritten.

nandojve replied:
Ok, so if there is no objections I would like to move forward this way. The developer can easy switch the peripheral clock input at board using:

```
&sercom0 {
	atmel,assigned-clocks = <&gclk 2>;
};
```
This way user can define manually other lines at board init and override the default configuration. I can send another proposal to complete this and auto configure the gclk at SoC init on another PR. In general, this can be improved with #72102 later.

nandojve replied:
Hi @danieldegrasse , @henrikbrixandersen ,

I was wondering if you had time to check my last comments and see last update.

henrikbrixandersen replied:
The approach looks sane to me, but I expect/fear we will have to rework it all once #72102 goes in.

Also, still not a fan of the odd formatting used for register assignments as already pointed out here: https://github.com/zephyrproject-rtos/zephyr/pull/82183#discussion_r1864458150

nandojve replied:
I can fix the `;`
We can always rework to make it better.

nandojve commented:
* rebase
* more context in commit messages

nandojve committed:
clocks: atmel: sam0: Fix gclk and mclk clock bindings

The Atmel SAM0 SoC enable peripherals clocks in distinct places: PM and
MCLK. The old devices had defined the peripheral clock enable bit at PM.
On the newer devices this was extracted on a dedicated memory section
called Master Clock (MCLK). This change excludes the dedicated bindings
in favor of a generic approach that cover all cases.

Now the clocks properties is complemented by the atmel,assigned-clocks
property. It gives the liberty to user to customize the clock source
from a generic clock or configure the direct connections.

All peripherals drivers were reworked with the newer solution.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
* rebase due to #82991 got merged.

nandojve commented:
rebase with latest fixes.

danieldegrasse reviewed (approved):
Ok with this given that it is all vendor specific
